Added documentation for Otel metrics (#3040)
elastic/apm-agent-java: docs/api-opentelemetry.asciidoc

=== OpenTelemetry bridge

The Elastic APM OpenTelemetry bridge allows creating Elastic APM `Transactions` and `Spans`,

> using the OpenTelemetry API. OpenTelemetry metrics are collected and shipped to elastic as well.

Reviewer: ```suggestion
using the OpenTelemetry API. OpenTelemetry metrics are also collected.
```

`elastic` in itself is a bit vague destination, and we'd better avoid any confusion where the metrics are sent.

In both cases the Elastic APM Agent will respect the <<config-disable-metrics, `disable_metrics`>> and <<config-metrics-interval, `metrics_interval`>> settings for OpenTelemetry metrics.

You can use the <<config-custom-metrics-histogram-boundaries, `custom_metrics_histogram_boundaries`>> setting to customize histogram bucket boundaries.
Alternatively you can use OpenTelemetry `Views` to define histogram buckets on a per-metric basis when providing your own `MeterProvider`.

[float]
[[otel-metrics-api]]
==== API Usage

You can define metrics and report metric data via `GlobalOpenTelemetry`:

[source,java]
----
import io.opentelemetry.api.GlobalOpenTelemetry;
import io.opentelemetry.api.metrics.LongCounter;
import io.opentelemetry.api.metrics.Meter;

Meter myMeter = GlobalOpenTelemetry.getMeter("my_meter");
LongCounter counter = meter.counterBuilder("my_counter").build();
counter.add(42);
----

We don't require you to setup an OpenTelemetry `MeterProvider` within `GlobalOpenTelemetry` yourself.
The Elastic APM Java Agent will detect if no `MeterProvider` was configured and will provide its own automatically in this case.

> If you provide your own `MeterProvider` (see <<otel-metrics-sdk>>), it will not be overridden by the agent.

Reviewer: ```suggestion
If you provide your own `MeterProvider` (see <<otel-metrics-sdk>>), the agent will use the provided instance.
```

[float]
[[otel-metrics-sdk]]
==== Using a customized MeterProvider

In some cases using just the <<otel-metrics-api, OpenTelemetry API for metrics>> might not be flexible enough.
Some example use cases are:

>  * You want to use OpenTelemetry Views (e.g. to customize histogram buckets on a per-metric basis)
>  * You want to export metrics to other tools in addition to Elastic APM (e.g. prometheus)

Reviewer: ```suggestion
 * Using OpenTelemetry Views (e.g. to customize histogram buckets on a per-metric basis)
 * Exporting metrics to other tools in addition to Elastic APM (e.g. prometheus)
```

For these use cases you can just setup you OpenTelemetry SDK `MeterProvider`.
The Elastic APM Agent will take care of installing an additional `MetricExporter` via instrumentation,
which will ship the metric data to Elastic APM.

> This requires you to use OpenTelemetry version `1.16.0` or newer.

Reviewer: ```suggestion
This requires using OpenTelemetry version `1.16.0` or newer.
```
[question] : is the version constraint from our agent instrumentation or something that is not supported by older OTel versions ? If it is the latter, then I don't think it's worth mentioning it here (as it is already implicitly covered in the Otel docs).

Author: Otel 1.16.0 added the capability for exporters to define the default aggregations to use (and temporality if I remember correctly). These features are used by our Otel Metric Exporter. For this reason the instrumentation ignores otel-SDKs which are older than 1.16.
For the API-based usage we support all opentelemetry versions (metrics were added in 1.10 )

[source,java]
----
import io.opentelemetry.sdk.metrics.SdkMeterProvider;
import io.opentelemetry.api.metrics.DoubleHistogram;
import io.opentelemetry.api.metrics.Meter;
import io.opentelemetry.api.metrics.MeterProvider;
import io.opentelemetry.exporter.prometheus.PrometheusHttpServer;
import io.opentelemetry.sdk.metrics.InstrumentSelector;
import io.opentelemetry.sdk.metrics.View;

//Elastic APM MetricReader will be registered automatically by the agent
SdkMeterProvider meterProvider = SdkMeterProvider.builder()
    .registerMetricReader(PrometheusHttpServer.create())
    .registerView(
        InstrumentSelector.builder().setName("my_histogram").build(),
        View.builder().setAggregation(Aggregation.explicitBucketHistogram(List.of(1.0, 5.0))).build()
    )
    .build();

Meter testMeter = meterProvider.get("my_meter");
DoubleHistogram my_histogram = testMeter.histogramBuilder("my_histogram").build();

my_histogram.record(0.5);
----

[float]
[[otel-caveats]]
==== Caveats
Not all features of the OpenTelemetry API are supported.

[float]
[[otel-metrics]]
===== Metrics
This bridge only supports the tracing API.
The Metrics API is currently not supported.

> 

Reviewer: :heart_eyes:
